src/Symfony/Component/Serializer/Normalizer/BackedEnumNormalizer.php

```diff
@@ -69,14 +69,17 @@ public function denormalize(mixed $data, string $type, string $format = null, ar
             }
         }
 
-        if (!\is_int($data) && !\is_string($data)) {
-            throw NotNormalizableValueException::createForUnexpectedDataType('The data is neither an integer nor a string, you should pass an integer or a string that can be parsed as an enumeration case of type '.$type.'.', $data, [Type::BUILTIN_TYPE_INT, Type::BUILTIN_TYPE_STRING], $context['deserialization_path'] ?? null, true);
+        $backingType = (new ReflectionEnum($type))->getBackingType()->getName();
+        $givenType = get_debug_type($data);
+
+        if ($givenType !== $backingType) {
+            throw NotNormalizableValueException::createForUnexpectedDataType(sprintf('Data expected to be "%s", "%s" given. You should pass a value that can be parsed as an enumeration case of type %s.', $backingType, $givenType, $type), $data, [$backingType], $context['deserialization_path'] ?? null, true);
         }
 
         try {
             return $type::from($data);
         } catch (\ValueError $e) {
-            throw new InvalidArgumentException('The data must belong to a backed enumeration of type '.$type);
+            throw NotNormalizableValueException::createForUnexpectedDataType('The data must belong to a backed enumeration of type '.$type, $data, [$backingType], $context['deserialization_path'] ?? null, true);
```

Comment:
this change basically reverts #47128
see #50192 also

Reply:
https://github.com/symfony/symfony/pull/47128 was resolving https://github.com/symfony/symfony/issues/46977 issue which is covered by the `\Symfony\Component\Serializer\Tests\SerializerTest::testCollectDenormalizationErrorsWithWrongEnum` test and it works as expected. It doesn't throw TypeError but gives PartialDenormalizationException as expected.

Reply:
Throwing `InvalidArgumentException` instead of `NotNormalizableValueException` undoes the feature done by https://github.com/symfony/symfony/pull/42502

I first viewed https://github.com/symfony/symfony/pull/43047#issuecomment-921077295
Then I was happy that may be https://github.com/symfony/symfony/pull/40830 + https://github.com/symfony/symfony/pull/42502 would be a solution to my problem. But then https://github.com/symfony/symfony/pull/47128 undid that functionality again. It was partially reverted back to an `NotNormalizableValueException` in https://github.com/symfony/symfony/pull/50192

The whole point of having it as an `NotNormalizableValueException` is that you should be able to serialize DTOs with enums and then collect all the errors at the end in an `PartialDenormalizationException`. You should take care not to ruin that functionality.